File: hackerrank/problem solving/algorithms/2 - implementation/7 - divisible sum pairs/divisible_sum_pairs.java
// https://www.hackerrank.com/challenges/divisible-sum-pairs/problem?isFullScreen=true
// Java 25

void main() {
    int k = readNumbers()[1];
    int[] numbers = readNumbers();
    IO.println(divisibleSumPairs(numbers, k));
}

int[] readNumbers() {
    return Stream.of(IO.readln().split(" ")).mapToInt(Integer::parseInt).toArray();
}

// n: length of numbers
// k: length of freq
// T: O(n + k)
// S: O(k) extra space
int divisibleSumPairs(int[] numbers, int k) {
    int[] freq = initRemainderFrequency(numbers, k);
    return countPairsWithRemainder0(freq) + countComplementaryRemainderPairs(freq) +
            countPairsWithRemainderKHalf(freq);
}

int[] initRemainderFrequency(int[] numbers, int k) {
    int[] freq = new int[k];
    for (int x : numbers)
        freq[x % k]++;
    return freq;
}

int countPairsWithRemainder0(int[] freq) {
    return pairCount(freq[0]);
}

int pairCount(int n) {
    return n * (n - 1) / 2;
}

int countComplementaryRemainderPairs(int[] freq) {
    int n = freq.length;
    return IntStream.range(1, (n + 1) / 2)
            .map(i -> freq[i] * freq[n - i])
            .sum();
}

int countPairsWithRemainderKHalf(int[] freq) {
    int n = freq.length;
    return n % 2 == 0 ? pairCount(freq[n / 2]) : 0;
}

File: hackerrank/problem solving/algorithms/2 - implementation/7 - divisible sum pairs/divisible_sum_pairs.go
// https://www.hackerrank.com/challenges/divisible-sum-pairs/problem?isFullScreen=true

package main

import "fmt"

func main() {
    var n, k int
    _, _ = fmt.Scan(&n, &k)
    numbers := readNumbers(n)
    fmt.Println(divisibleSumPairs(numbers, k))
}

func readNumbers(n int) []int {
    arr := make([]int, n)
    for i := range n {
        _, _ = fmt.Scan(&arr[i])
    }
    return arr
}

// n: length of numbers
// k: length of freq
// T: O(n + k)
// S: O(k) extra space
func divisibleSumPairs(numbers []int, k int) int {
    freq := initRemainderFrequency(numbers, k)
    return countPairsWithRemainder0(freq) + countComplementaryRemainderPairs(freq) + countPairsWithRemainderKHalf(freq)
}

func initRemainderFrequency(numbers []int, k int) []int {
    freq := make([]int, k)
    for _, x := range numbers {
        freq[x%k]++
    }
    return freq
}

func countPairsWithRemainder0(freq []int) int {
    return pairCount(freq[0])
}

func pairCount(n int) int {
    return n * (n - 1) / 2
}

func countComplementaryRemainderPairs(freq []int) int {
    count := 0
    n := len(freq)
    for i, n2 := 1, (n+1)/2; i < n2; i++ {
        count += freq[i] * freq[n-i]
    }
    return count
}

func countPairsWithRemainderKHalf(freq []int) int {
    n := len(freq)
    if n%2 == 0 {
        return pairCount(freq[n/2])
    }
    return 0
}
